File: ios/FoodHomeApp/Notifications/PushLifecyclePolicy.swift
import CryptoKit
import Foundation

enum NotificationAuthorizationStatus: String, Equatable {
    case notDetermined
    case denied
    case authorized
    case provisional
    case unavailable
}

struct PushOpen: Equatable {
    let eventID: String
    let route: URL
}

struct PushPayloadPolicy {
    let navigationPolicy: NavigationPolicy

    func parse(_ payload: [String: String]) -> PushOpen? {
        let allowedKeys: Set<String> = ["eventId", "route", "type"]
        guard Set(payload.keys).isSubset(of: allowedKeys),
              let eventID = payload["eventId"],
              eventID.range(
                of: "^[A-Za-z0-9][A-Za-z0-9._:-]{0,63}$",
                options: .regularExpression
              ) != nil,
              let route = payload["route"],
              case let .internal(trustedURL) = navigationPolicy.classify(route)
        else {
            return nil
        }
        return PushOpen(eventID: eventID, route: trustedURL)
    }
}

final class SensitivePushToken: CustomStringConvertible {
    private let rawValue: Data
    let fingerprint: String

    init?(_ rawValue: Data) {
        guard !rawValue.isEmpty, rawValue.count <= 4_096 else { return nil }
        self.rawValue = rawValue
        self.fingerprint = SHA256.hash(data: rawValue)
            .prefix(6)
            .map { String(format: "%02x", $0) }
            .joined()
    }

    func use<T>(_ block: (Data) -> T) -> T {
        block(rawValue)
    }

    var description: String {
        "SensitivePushToken(<redacted>:\(fingerprint))"
    }
}

File: android/app/src/main/java/market/foodhome/app/notifications/PushLifecyclePolicy.kt
package market.foodhome.app.notifications

import market.foodhome.app.navigation.NavigationDecision
import market.foodhome.app.navigation.NavigationPolicy
import java.security.MessageDigest

enum class NotificationAuthorizationStatus(val wireValue: String) {
    NotDetermined("notDetermined"),
    Denied("denied"),
    Authorized("authorized"),
    Provisional("provisional"),
    Unavailable("unavailable"),
}

data class PushOpen(
    val eventId: String,
    val route: String,
)

class PushPayloadPolicy(
    private val navigationPolicy: NavigationPolicy,
) {
    private val eventIdPattern = Regex("^[A-Za-z0-9][A-Za-z0-9._:-]{0,63}$")

    fun parse(payload: Map<String, String>): PushOpen? {
        if (payload.keys.any { it !in ALLOWED_KEYS }) return null
        val eventId = payload["eventId"]?.takeIf(eventIdPattern::matches) ?: return null
        val route = payload["route"] ?: return null
        val decision = navigationPolicy.classify(route) as? NavigationDecision.Internal ?: return null
        return PushOpen(eventId, decision.uri.toASCIIString())
    }

    private companion object {
        val ALLOWED_KEYS = setOf("eventId", "route", "type")
    }
}

class SensitivePushToken private constructor(
    private val rawValue: String,
) {
    val fingerprint: String = MessageDigest.getInstance("SHA-256")
        .digest(rawValue.toByteArray(Charsets.UTF_8))
        .take(6)
        .joinToString("") { "%02x".format(it.toInt() and 0xff) }

    fun <T> use(block: (String) -> T): T = block(rawValue)

    override fun toString(): String = "SensitivePushToken(<redacted>:$fingerprint)"

    companion object {
        fun from(rawValue: String): SensitivePushToken? = rawValue
            .takeIf { it.isNotBlank() && it.length <= 4_096 }
            ?.let(::SensitivePushToken)
    }
}

fun interface PushTokenSink {
    fun receive(token: SensitivePushToken)
}

object DisabledPushTokenSink : PushTokenSink {
    override fun receive(token: SensitivePushToken) {
        // Intentionally disabled until the accepted MobileInstallation binding contract exists.
    }
}
